### PR TITLE
Use umbrella crate for minimal template

### PR DESCRIPTION
**Pull request by pgherveou**

### CODE REVIEW

File: templates/minimal/node/Cargo.toml
Comment (franciscoaguirre) [inline templates/minimal/node/Cargo.toml]: 🎉

**Comment by ggwpez** [comment]
Some of the clippy warnings seem to come from the `experimental` feature...

**Comment by pgherveou** [comment]
> Some of the clippy warnings seem to come from the `experimental` feature...

Yeah this is weird as the experimental feature seems to be propagated correctly

File: templates/minimal/pallets/template/src/lib.rs
Comment (bkchr) [inline templates/minimal/pallets/template/src/lib.rs:18]:
```suggestion
```

Comment (pgherveou) [reply templates/minimal/pallets/template/src/lib.rs:18]: Right let me fix 
https://github.com/paritytech/polkadot-sdk/blob/d3d1542c1d387408c141f9a1a8168e32435a4be9/substrate/frame/support/procedural/src/pallet/parse/config.rs?plain=1#L274
instead

File: templates/minimal/runtime/src/lib.rs
Comment (kianenigma) [inline templates/minimal/runtime/src/lib.rs:32]: The whole value of `polkadot-sdk-frame` is in using `prelde::*`, please see https://paritytech.github.io/polkadot-sdk/master/polkadot_sdk_frame/index.html

**Review by kianenigma — CHANGES_REQUESTED** [review]
I think we are waiting for a new release before we can merge this? 

also would like to see https://github.com/paritytech/polkadot-sdk/pull/5155/files#r1703858118 addressed.

**Review by ggwpez — APPROVED** [review]
Thanks for the demo!

File: templates/minimal/node/Cargo.toml
Comment (kianenigma) [inline templates/minimal/node/Cargo.toml:28]: why is this feature flag specifically needed?

File: templates/minimal/runtime/Cargo.toml
Comment (kianenigma) [inline templates/minimal/runtime/Cargo.toml:30]: This might be missing in `frame`, you should add it if so.

**Review by kianenigma — APPROVED** [review]
Made a few fixes me self now, and now it is all good!

**Comment by pgherveou** [comment]
👀  thanks for babysitting this PR, just came back from vacations will take a look at this

**Review by EgorPopelyaev — APPROVED** [review]

**Comment by kianenigma** [comment]
must admit the CI in this PR has just been a terrible user experience, constantly getting unrelated failures :(

**Comment by pgherveou** [comment]
We probably need to update the rve-nightly to use 1.77 to fix the compilation bug
the last commit is just an attempt to check if that fixes the build.

I have a PR to update the rve toolchain here https://github.com/paritytech/rustc-rv32e-toolchain/pull/35

**Comment by kianenigma** [comment]
You are an absolute champion for fighting this CI @pgherveou

File: substrate/frame/revive/src/wasm/runtime.rs
Comment (athei) [inline substrate/frame/revive/src/wasm/runtime.rs:559]: Merge error? This shouldn't be necessary?

Comment (pgherveou) [reply substrate/frame/revive/src/wasm/runtime.rs:559]: I think it is when this jobs run
```
SUBSTRATE_RUNTIME_TARGET=riscv cargo check -p minimal-template-runtime
```

Comment (athei) [reply substrate/frame/revive/src/wasm/runtime.rs:559]: Curious. I guess due to the difference in toolchain version then.

Comment (pgherveou) [reply substrate/frame/revive/src/wasm/runtime.rs:559]: so it does now when we build 
```
SUBSTRATE_RUNTIME_TARGET=riscv cargo check -p minimal-template-runtime
```
because it pulls the umbrella crate that tries to build all the crates deps including pallet-revive to riscv.

Comment (pgherveou) [reply substrate/frame/revive/src/wasm/runtime.rs:559]: @kianenigma that explains why we are seeing all these compilations errors here and not in other PRs.
We could pick only the features we need in the Cargo.toml if we want to avoid that ...

Comment (ggwpez) [reply substrate/frame/revive/src/wasm/runtime.rs:559]: Hm, yea only selecting the needed dependencies should work.  
It's a bit annoying though, since the umbrella crate is supposed to make things easier. I guess it is still easier than selecting each dependency version manually.

Would it be possible to feature-gate the RISCV compilation?

Comment (pgherveou) [reply substrate/frame/revive/src/wasm/runtime.rs:559]: I gated the pallets that were causing issue, the long term solution is probably to generate the umbrella trait with feature flags that declare all their dependencies, so that you can hand pick the one you need and optimize your build time

Comment (athei) [reply substrate/frame/revive/src/wasm/runtime.rs:559]: > yea only selecting the needed dependencies should work.
> It's a bit annoying though, since the umbrella crate is supposed to make things easier.

Couldn't we encode the dependency graph into the features in the script? Like you only activate the pallets you want and then they automatically activate the other pallets they depend on.

Comment (pgherveou) [inline substrate/frame/revive/src/wasm/runtime.rs:559]:
```suggestion
impl<'a, E: Ext, M: ?Sized + Memory<E::T>> Runtime<'a, E, M> {
```

rolling back

File: substrate/frame/support/test/tests/pallet_outer_enums_explicit.rs
Comment (pgherveou) [inline substrate/frame/support/test/tests/pallet_outer_enums_explicit.rs:1]: note to self: maybe this can be rolled back as well

File: substrate/frame/revive/src/wasm/runtime.rs
Comment (kianenigma) [reply substrate/frame/revive/src/wasm/runtime.rs]: I would not support making the usage of umbrella crate in any of the templates sub-optimal, for the sake of `SUBSTRATE_RUNTIME_TARGET=riscv` which is more of an experimental thing. 

All people out there will not face the same compilation issue, and people do use these templates as a source of truth, so it should be in the simplest form possible.

File: templates/minimal/runtime/Cargo.toml
Comment (kianenigma) [inline templates/minimal/runtime/Cargo.toml:15]: This is what I mean, is there really no way to avoid this? esp in our 3 templates, I don't care if e.g. AH that will have pallet-revive are built later like this.

Comment (kianenigma) [reply templates/minimal/runtime/Cargo.toml:15]: I was aware of the tradeoff and was leaning towards the fact that most novice ppl prefer a "one-time" over compilation, in return for never having to touch their toml file again, and letting their IDK figure out which trait comes in from which sp- crate. 

I am not against the runtime feature, but still mildly prefer it to not be the norm. But I'll leave that to the wisdom of the crowd. 

By the way, polkadot-sdk/runtime is meant to be just the things that are needed for building a runtime, minus pallet, it would be more or less identical to frame/runtime.

Comment (athei) [reply templates/minimal/runtime/Cargo.toml:15]: I think it is reasonable to ask people to add a feature when adding a new pallet. It is not something you do regularly after setting up your runtime. The compilation cost you pay all the time. We experienced first hand just now how including `polkadot_sdk` into `revive` balooned the compilation time of that project.

**Review by athei — APPROVED** [review]
I think with this we kill two birds with one stone. Cut down on compile times for our users and fix our CI.

**Review by serban300 — APPROVED** [review]

**Pull request merged**
